Xfstests tool & test update to support Azure file Share - SMB testing

shekharsorot opened the pull request:
Xfstest is a file system and share testing tool that is maintained by kernel.org and is often used for performance, stability, feature operability and regression testing.

This LISA tool feature is an update to the existing Xfstests.py within Microsoft test cases and has been refactored to be able to run SMB as well as NFS tests in addition of existing set of Managed disks ( XFS, EXT3/4 and BTRFS ) as well as NVME

It's an implementation of "https://git.kernel.org/pub/scm/fs/xfs/xfstests-dev.git", although users can specify custom repository and forks as well with their own test cases.

It's intended to use this via custom test suites to test SMB and NFS client functionality with various SMB and NFS servers (in addition to AZFiles ) currently in use across cloud and on-premises environments (You can call XFStest to create config file for existing shares, specify custom flags in addition to existing flags and call run_test() to trigger testing // reporting.

Note: detailed docstrings are included in tool class and respective method definitions.

Non-Breaking but Major changes:

TIMEOUT changes to 6 hours to allow additional time for managed disk tests to run.
SMB // CIFS test is hardcoded for specific cases as these are well known tests for Azure Files - SMB Premium // Standard
XFStests will now always pull the 'master' branch which is the last known stable release.
"verify_azure_file_share" test case priority has been reduced back to P3 as the test takes < 60 min to complete for CIFS
Xfstests will by default generate dmesg output per case. This can be found in your runtime / log / <Test Project > /testss / xfstests directory for analysis. 
SubtestMessages will now contain diff of input vs output file for failed cases in the additional information field. Can be used for automated reporting when test environment is not preserved.
Upon Failure of any sub test, HTML will output only the failed subtests.
In case of success ( including skipped  and expunged ), HTML report will dump the list of passed test cases along with test options.

Usage: 
import using "from lisa.testsuite import TestResult"

Instantiate & install using "xfstests = node.tools[Xfstests]"

to create a config:
xfstest.set_local_config(
            scratch_dev="/dev/sdb",
            scratch_mnt="/mnt/scratch",
            test_dev="/dev/sdc",
            test_folder="/mnt/test",
            file_system="xfs",
            test_section="xfs-custom",
            mount_opts="noatime",
            testfs_mount_opts="noatime",
            additional_parameters={"TEST_DEV2": "/dev/sdd"},
            overwrite_config=True
            )
Note: You can create multiple test sections using this method and call run_test() in sequence.
Please note guidance in "https://github.com/kdave/xfstests/blob/master/README.config-sections"


To exclude global list of tests
xfstests.set_excluded_tests(excluded_tests)


To run tests
xfstests.run_test(
            test_section=test_type,
            log_path=log_path,
            result=result,
            data_disk=data_disk,
            test_cases=test_cases,
            timeout=self.TIME_OUT - 30,
        )

Tool instance will create and send sub-test messages to log as well as HTML & Junit.
On failure, it will display the failed sub-test as well as XFStest message to allow for troubleshooting.
On success, it will display list of passed // skipped sub-tests via PassedException.
Note: Expunged cases are treated as skipped
        
Supported OS: 
redhat, suse , fedora, mariner, debian

shekharsorot committed:
add XFStests as native LISA tool

shekharsorot committed:
Update xfstests.py

shekharsorot committed:
Update xfstests.py

shekharsorot committed:
Update xfstests.py

shekharsorot committed:
Update xfstests.py

shekharsorot committed:
Update xfstests.py

shekharsorot committed:
Update xfstests.py

shekharsorot committed:
Update xfstests.py

shekharsorot committed:
Update xfstests.py

squirrelsc reviewed (commented):
> I will work to update the XFStesting suite in Microsoft folder. It will take some time due to long run durations and validations for each test case. I will modify xfstesting.py to use the new Xfstests tool. 

There is a xfstests.py file already, can you directly fix there?

shekharsorot committed:
update XFSTesting and Xfstests tool

shekharsorot committed:
Update xfstesting.py

shekharsorot committed:
Update xfstesting.py

shekharsorot committed:
Update xfstesting.py

shekharsorot committed:
Update xfstesting.py

shekharsorot committed:
Update xfstesting.py

shekharsorot committed:
Update xfstesting.py

shekharsorot committed:
Update xfstests.py

shekharsorot committed:
Update xfstesting.py

shekharsorot committed:
Update xfstesting.py

shekharsorot committed:
fixes and optimizations

shekharsorot committed:
fix for long running disk tests

shekharsorot committed:
syntax fix for CIFS, forcing test group for all other cases

shekharsorot committed:
Update xfstesting.py

shekharsorot committed:
performance fixes for diagnostics message generation

shekharsorot committed:
move generic 586 for Azfiles to excluded list

shekharsorot committed:
reducing timeout back to 4 hours

reducing timeout to 4 hours as original since we no longer are seeing issues with quick tests running > 2 hours with new XFStests code

shekharsorot committed:
fixed issue of >6 return statements in create_xfstest_stack_info()

shekharsorot committed:
remove type_checking for xfstests.py tool

shekharsorot committed:
add note for device ID for mount points

shekharsorot committed:
Flake8 fix and AZfiles test priority change

shekharsorot committed:
fix for Azure files SMB multi channel + share def changes to accept list of share names

shekharsorot committed:
Update xfstests.py

shekharsorot committed:
Update xfstesting.py

shekharsorot committed:
Minor bug fix due to un-intended except block not correctly reflecting correct test status on console.

shekharsorot committed:
Update xfstests.py

shekharsorot committed:
reverted verify_azure_file_share ti not using try and finally block, relying on LISA for resource life cycle management

shekharsorot committed:
added comments to xfstst.py for create_xfstest_stack_info() method
